### PR TITLE
[Do Not Merge]Inital cugraph support 

### PR DESCRIPTION
**Pull request by VibhuJawa**
## Description

This PR add cugraph support to dgl by adding a `CuGraphStorage` which follows the API convention described [here](https://hackmd.io/6PJixJj7Ssy19I-N9CV1Ag?view)  to enable use of DGL DataLoader and certain features, including various neighborhood samplers, negative samplers, and feature prefetching,

This work can be extended to `GaasGraphStorage`  which would enable sampling with multiple GPUs . 

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes
- [x] Add  CuGraphStorage class
- [x] Add  tests around CuGraphStorage
- [ ] Add unit tests around other dataloader parameters for cugraph sampling
- [x] Add single gpu training example with CuGraphStorage
- [ ] Add multi gpu training example with CuGraphStorage

### CODE REVIEW

**Comment by BarclayII** [comment]
I just figured out that the example seems to be based on the old example.  Could you rewrite the example to something like `examples/pytorch/graphsage/node_classification.py`?  There should be one single script with both the model and training logic, which is much simpler than the old examples.

File: python/dgl/contrib/cugraph/cugraph_storage.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        # change the seed_nodes from pytorch tensor to cudf series
+        if torch.is_tensor(seed_nodes):
+            seed_nodes = to_dlpack(seed_nodes)
+            seed_nodes = cudf.from_dlpack(seed_nodes)
+
+        parents_nodes, children_nodes = self.graphstore.sample_neighbors(
+            seed_nodes, fanout, edge_dir="in", prob=None, replace=False
+        )
+        # FIXME: why are children_nodes src and parents_nodes dst here
+        sample_df = cudf.DataFrame({"_SRC_": children_nodes, "_DST_": parents_nodes})
```
Comment (BarclayII) [inline python/dgl/contrib/cugraph/cugraph_storage.py:137]: Does "children" mean the nodes out of which the edges connect, and "parent" mean the nodes in which the edges connect?

File: python/dgl/contrib/cugraph/cugraph_storage.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        del parents_nodes, children_nodes
```
Comment (BarclayII) [inline python/dgl/contrib/cugraph/cugraph_storage.py:138]: Why do we need to delete them explicitly?  I think the peak memory consumption will be the same regardless of whether you delete them.

File: python/dgl/contrib/cugraph/cugraph_storage.py [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+            If True, the extracted subgraph will only have the nodes in the
+            specified node set and it will relabel the nodes in order.
+        output_device : Framework-specific device context object, optional
+            The output device.  Default is the same as the input graph.
+        Returns
+        -------
+        DGLGraph
+            The subgraph.
+        """
+        pass
```
Comment (BarclayII) [inline python/dgl/contrib/cugraph/cugraph_storage.py:224]: Could raise a `NotImplementedError`.

File: python/dgl/contrib/cugraph/cugraph_storage.py [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+        return src_nodes_tensor, dst_nodes_tensor
+
+    # Required in Link Prediction negative sampler
+    def num_nodes(self, ntype=None):
+        """Return the number of nodes for the given node type."""
+        # use graphstore function
+        if ntype:
+            return self._ndata[self._ndata["_TYPE_"] == ntype].shape[0]
+        else:
+             return self._ndata.shape[0]
```
Comment (BarclayII) [inline python/dgl/contrib/cugraph/cugraph_storage.py:247]: Indentation doesn't look correct.

File: python/dgl/contrib/cugraph/cugraph_storage.py [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+    def sample_neighbors(
+        self,
+        seed_nodes,
+        fanout,
+        edge_dir="in",
+        prob=None,
+        exclude_edges=None,
+        replace=False,
+        output_device=None,
+    ):
```
Comment (BarclayII) [inline python/dgl/contrib/cugraph/cugraph_storage.py:84]: I don't think closing parenthesis on a single line will pass lint checks.

**Review by BarclayII — CHANGES_REQUESTED** [review]

File: python/dgl/contrib/cugraph/cugraph_utils.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+# limitations under the License.
+
+import cugraph
+import cudf
+import dgl
+import cupy
+
+
+# from cugraph to DGL using
+def cugraphToDGL(graph):
```
Comment (jermainewang) [inline python/dgl/contrib/cugraph/cugraph_utils.py:21]: Let's have a separate PR to add this function to DGL. You could put the function in `dgl/convert.py`. Suggested signature:
```python
def from_cugraph(cug):
    """Create a graph from a cuGraph graph object.

    Parameters
    ----------
    cug : cugraph.Graph
        The input cuGraph object.

    Returns
    -------
    g_dgl : DGLGraph
        The created DGL graph object.
    """
```

File: python/dgl/contrib/cugraph/cugraph_utils.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    g_dgl : DGLGraph
+    """
+    edgelist = graph.edges()
+    src = cupy.asarray(edgelist["src"])
+    dst = cupy.asarray(edgelist["dst"])
+    g_dgl = dgl.graph((src, dst))
+    return g_dgl
+
+
+def dglToCugraph(graph):
```
Comment (jermainewang) [inline python/dgl/contrib/cugraph/cugraph_utils.py:40]: Similarly, let's have this function in DGL. You could put the function in `dgl/convert.py`. Suggested signature:
```python
def to_cugraph(g):
    """Convert a DGL graph to cuGraph object.
    
    The function only converts graph structure while node/edge features are simply dropped.
    Only support homogeneous graph input.

    Parameters
    -----------
    g : DGLGraph
        The input DGL graph.

    Returns
    -------
    cug : cugraph.Graph
        The converted cuGraph object.
    """
```

Comment (VibhuJawa) [reply python/dgl/contrib/cugraph/cugraph_utils.py:40]: Thanks for the review, Will create a separate PR for this.

File: python/dgl/contrib/cugraph/cugraph_utils.py (hunk 6 above)
Comment (VibhuJawa) [reply python/dgl/contrib/cugraph/cugraph_utils.py:21]: Thanks for the review, Will create a separate PR for this.

Comment (VibhuJawa) [reply python/dgl/contrib/cugraph/cugraph_utils.py:21]: Started a PR for this https://github.com/dmlc/dgl/pull/4166 here.

File: python/dgl/contrib/cugraph/cugraph_utils.py (hunk 7 above)
Comment (VibhuJawa) [reply python/dgl/contrib/cugraph/cugraph_utils.py:40]: Started a PR for this https://github.com/dmlc/dgl/pull/4166 here.

**Comment by VibhuJawa** [comment]
@jermainewang , This PR will miss this release.  The cugraph changes seems to be more involved that i initially guessed  . Please feel free  to remove the `Release Candidate` tag from the PR.

**Pull request merged**